include/picongpu/plugins/Emittance.hpp

```diff
@@ -0,0 +1,989 @@
+#include <pmacc/traits/HasFlag.hpp>
+#include <pmacc/algorithms/ForEach.hpp>
+#include <pmacc/math/Vector.hpp>
+
+#include <boost/mpl/and.hpp>
+
+#include <vector>
+#include <algorithm>
+#include <utility>
+#include <string>
+#include <iostream>
+#include <fstream>
+#include <memory>
+#include <stdexcept>
+
+
+
+
+namespace picongpu
+{
+
+    /** calculates the emittance in x direction
+     */
+    template< uint32_t T_numWorkers >
+    struct KernelCalcEmittance
+    {
+
+        /** calculates the sum of x^2, ux^2 and (x*ux)^2 and counts electrons
+         *
+         * @tparam T_ParBox pmacc::ParticlesBox, particle box type
+         * @tparam T_DBox pmacc::DataBox, type of the memory box for the reduced values
+         * @tparam T_Mapping mapper functor type
+         *
+         * @param pb particle memory
+         * @param gSum...,gCount_e storage for the reduced values
+         * @param mapper functor to map a block to a supercell
+         */
+        template<
+            typename T_ParBox,
+            typename T_DBox,
+            typename T_Mapping,
+            typename T_Acc,
+            typename T_Filter
+        >
+        DINLINE void operator( )(
+            T_Acc const & acc,
+            T_ParBox pb,
+            T_DBox gSumMom2,
+            T_DBox gSumPos2,
+            T_DBox gSumMomPos2,
+            T_DBox gCount_e,
+            DataSpace<simDim> globalOffset,
+            const int subGridY,
+            T_Mapping mapper,
+            T_Filter filter
+        ) const
+        {
+            using namespace mappings::threads;
+            constexpr uint32_t numWorkers = T_numWorkers;
+            constexpr uint32_t numParticlesPerFrame = pmacc::math::CT::volume<
+                typename T_ParBox::FrameType::SuperCellSize
+            >::type::value;
+
+            uint32_t const workerIdx = threadIdx.x;
+
+            using FramePtr = typename T_ParBox::FramePtr;
+
+            // shared sums of x^2, ux^2, (x*ux)^2, particle counter
+            PMACC_SMEM( acc, shSumMom2, memory::Array< float_X,SuperCellSize::y::value > );
+            PMACC_SMEM( acc, shSumPos2, memory::Array< float_X,SuperCellSize::y::value > );
+            PMACC_SMEM( acc, shSumMomPos2, memory::Array< float_X,SuperCellSize::y::value > );
+            PMACC_SMEM( acc, shCount_e, memory::Array< float_X,SuperCellSize::y::value > );
+
+            using ParticleDomCfg = IdxConfig<
+                numParticlesPerFrame,
+                numWorkers
+            >;
+
+            using SuperCellYDom = IdxConfig<
+                SuperCellSize::y::value,
+                numWorkers
+            >;
+
+
+            ForEachIdx< SuperCellYDom >{ workerIdx }(
+                [&](
+                    uint32_t const linearIdx,
+                    uint32_t const
+                )
+                {
+                    // set shared sums of x^2, ux^2, (x*ux)^2, particle counter to zero
+                    shSumMom2[linearIdx] = float_X(0.0);
```

Comment:
Since #2622 is merged, isn't it better to use literals with `_X` instead of `float_X` constants? E.g. `0.0_X` instead of `float_X(0.0)`.